ssl/t1_lib.c

```diff
@@ -1436,18 +1390,50 @@ SSL_TICKET_RETURN tls_decrypt_ticket(SSL *s, const unsigned char *etick,
         }
         *psess = sess;
         if (renew_ticket)
-            return SSL_TICKET_SUCCESS_RENEW;
+            ret = SSL_TICKET_SUCCESS_RENEW;
         else
-            return SSL_TICKET_SUCCESS;
+            ret = SSL_TICKET_SUCCESS;
+        goto end;
     }
     ERR_clear_error();
     /*
      * For session parse failure, indicate that we need to send a new ticket.
      */
-    return SSL_TICKET_NO_DECRYPT;
- err:
+    ret = SSL_TICKET_NO_DECRYPT;
+
+ end:
     EVP_CIPHER_CTX_free(ctx);
     HMAC_CTX_free(hctx);
+
+    /*
+     * If set, the decrypt_ticket_cb() is always called regardless of the
+     * return value determined above. The callback is responsible for checking
+     * |ret| before it performs any action
```

Comment:
The documentation looks rather misleading with respect to the above statement:
```
       dec_cb is the application defined callback invoked after session ticket
       decryption has been attempted and any session ticket application data
       is available.  The application can call
       SSL_SESSION_get_ticket_appdata() at this time to retrieve the
       application data. 
```
Sure sounds like the ticket is guaranteed decrypted when the callback is made. Indeed why make the callback if it is not?  The only hint of more interesting possibilities is much further down under `RETURN VALUES`:
```
       The dec_cb callback must return one of the following SSL_TICKET_RETURN
       values. Under normal circumstances the retv value is returned
       unmodified, but the callback can change the behavior of the post-ticket
       decryption code by returning something different. The dec_cb callback
       must check the retv value before performing any action.
```
The last sentence does not quite convey that the ticket might not be available...

Perhaps I'm misreading the docs, but I'm confused after the first pass.

Reply:
Yes, the intent was to know if/when a ticket was received, and what happened after attempted decryption. If no ticket extension is received then it should not be called (tickets not used in this case). If an empty ticket extension is received, then it should be called with the appropriate `retv` value. The documentation should be cleared up.

Reply:
Can you provide the *motivation* for this design choice? It rather invites applications that register for this callback to do apparently bad things, like modify error conditions into non-error conditions, or attempt to use the session pointer when no decryption took place and it is uninitialized, ...  Is there a good use-case for making the callback on failure?

Reply:
Yes, applications (I'm looking from CDNs) will want to know if a bad ticket has been received; they may want to abort the connection (rather than create a new session) and/or may want to not send a new ticket. CDNs also report on these conditions to track potential stale clients and/or DDOS. I can see that if the SSL_TICKET_FATAL_ERR_OTHER/MALLOC condition is sent to the function that it shouldn't be changed; and there may be one or two other conditions as well that should not occur. These can be checked after the callback to ensure compliance (or an assert of some type).

Reply:
First  and foremost, the documentation needs to clearly spell out the contract between OpenSSL and the callback author. When will the callback be called (or not called), what parameters it will see under various conditions, and what return values it may use to signal some desired outcome.

How, for example, might an application indicate that no new tickets should be issued if decryption fails?  Normally, that would just result in a full handshake with new tickets issued. I don't, for example, see any return codes that signal suppression of ticket issuance while continuing the handshake. And SSL_TICKET_ERR_OTHER is rather a mixture of symptoms, I don't see it as a good condition for applications to raise.

The lack of complete documentation indicates to me a good likelihood of an incomplete design. What should we do here?  Since these callbacks are new in 1.1.1, perhaps they should be removed and wait for the next release and a more complete design?

The commit that created the callback just said
```
commit df0fed9aab239e2e9a269d06637a6442051dee3b
Author: Todd Short <tshort@akamai.com>
Date:   Wed Mar 15 13:25:55 2017 -0400

    Session Ticket app data
    
    Adds application data into the encrypted session ticket
    
    Reviewed-by: Paul Dale <paul.dale@oracle.com>
    Reviewed-by: Matt Caswell <matt@openssl.org>
    (Merged from https://github.com/openssl/openssl/pull/3802)
```
Neither this comment, nor #3802 suggests a broader agenda of giving applications control of the ticket behaviour, all that's promised and reviewed is access to "appdata", for which the callback should only happen on success.

So my impression is that the non-success cases are underspecified, and perhaps not ready for prime-time???

Reply:
As the result of parsing the application data, the application may choose to use or discard the ticket; that was part of the intent. If that wasn't clear, I apologize. I'm more than willing to create a PR that updates the documentation and sanitizes the input/output of the callback.

Reply:
Something needs to be done, since the current contract is rather under-specified.  For example under "RETURN VALUES" only the *input* semantics of the decrypt callback `retv` are specified. There is no documentation of what output values may be specified which states are valid input states from which to return a given output...  There's no warning about when the `ss` parameter might be invalid and `SSL_SESSION_get0_ticket_appdata()` should not be called.

If the callback is to stay it needs complete documentation, from which we can assess whether the implementation matches the design contract.  If that's not possible to complete before the release, I'd like to see the feature delayed to a later release.

Reply:
I'd suggest the callback should not be called in the event of any of the "fatal" return codes occurring. It can be called in any other event but the callback should not attempt to turn an error return code into a success. It seems to me to be within the scope of this PR to do that, so I'll give that a go.

Reply:
In the case of FATAL errors, I generally agree (not strong enough thoughts to disagree). Once @mattcaswell updates the code, I will update the documentation to reflect reality.